ctl/ctCpuAffinity.cpp

```diff
@@ -0,0 +1,220 @@
+
+*/
+
+/*
+ * ctCpuAffinity.cpp
+ * Implementation of CPU affinity capability detection and shard -> affinity mapping.
+ */
+
+#include "ctCpuAffinity.hpp"
+
+#include <cstdio>
+
+#include <winsock2.h>
+#include <mstcpip.h>
+#include <windows.h>
+#include <vector>
+#include <algorithm>
+
+namespace ctl
+{
+
+    CpuAffinityInfo QueryCpuAffinitySupport() noexcept
+    {
+        CpuAffinityInfo info{};
+
+        // Processor group and counts
+        info.ProcessorGroupCount = static_cast<uint32_t>(GetActiveProcessorGroupCount());
+        // total logical processors across groups
+        uint64_t total = 0;
+        for (WORD g = 0; g < info.ProcessorGroupCount; ++g)
+        {
+            total += GetActiveProcessorCount(g);
+        }
+        info.LogicalProcessorCount = static_cast<uint32_t>(total);
+
+        // Probe SIO_CPU_AFFINITY support using a temporary UDP socket, if available
+        // WSAStartup is expected to have been called by the application earlier.
+#ifdef SIO_CPU_AFFINITY
+        SOCKET s = WSASocketW(AF_INET, SOCK_DGRAM, IPPROTO_UDP, nullptr, 0, WSA_FLAG_OVERLAPPED);
+        if (s != INVALID_SOCKET)
+        {
+            DWORD bytes = 0;
+            // Per MSDN, SIO_CPU_AFFINITY expects a USHORT processor index (0-based)
+            // as its input buffer. Probe support by passing a small USHORT buffer.
+            USHORT probeProcessorIndex = 0;
+            const int rc = WSAIoctl(s, SIO_CPU_AFFINITY, &probeProcessorIndex, sizeof(probeProcessorIndex), nullptr, 0, &bytes, nullptr, nullptr);
+            info.SupportsCpuAffinityIoctl = (rc == 0);
+            closesocket(s);
+        }
+        else
+        {
+            info.SupportsCpuAffinityIoctl = false;
+        }
+#else
+        // SDK doesn't define SIO_CPU_AFFINITY; treat as not supported.
+        info.SupportsCpuAffinityIoctl = false;
+#endif
+
+        // Feature-detect GetQueuedCompletionStatusEx presence conservatively
+        // Assume present on supported Windows versions; leave false only if very old.
+        info.SupportsGQCSEx = true;
+
+        return info;
+    }
+
+    namespace {
+        // Helper: get per-group processor counts
+        static std::vector<uint32_t> GetProcessorCountsPerGroup() noexcept
+        {
+            std::vector<uint32_t> counts;
+            const WORD groupCount = static_cast<WORD>(GetActiveProcessorGroupCount());
+            counts.reserve(groupCount);
+            for (WORD g = 0; g < groupCount; ++g)
+            {
+                counts.push_back(static_cast<uint32_t>(GetActiveProcessorCount(g)));
+            }
+            return counts;
+        }
+
+        // Convert a global cpu index to group and local index
+        static void GlobalCpuIndexToGroupAndIndex(uint32_t globalIndex, const std::vector<uint32_t>& perGroupCounts, WORD& outGroup, uint32_t& outIndex) noexcept
+        {
+            uint32_t acc = 0;
+            for (WORD g = 0; g < perGroupCounts.size(); ++g)
```

Comment:
which it shouldn't happen in practice, we should fail if perGroupCounts.size() > MAX_WORD -- else this would overflow 'g'.